### PR TITLE
Feature persistor

### PR DESCRIPTION
**Pull request by cratcliff**
* Includes target probability and over all years.
* Moved stack and clip from k-means cluster into raster_ops.py for use by k-means and persistor.
Travis CI still fails.- Will look int this later but probably due to a package update.

### CODE REVIEW

File: pyprecag/tests/test_bandops.py [hunk 1]
```diff
@@ -42,40 +42,54 @@ def tearDown(self):
 
     def test_band_mapping(self):
         with self.assertRaises(AttributeError) as msg:
-            bm = BandMapping(red=4, green=3,blud=2)
-            self.assertEqual("BandMapping has no attribute blud. Allowed keys are blue, mask, infrared, rededge, green, red", str(msg.exception))
+            bm = BandMapping(red=4, green=3, blud=2)
```
Comment (dgobbett) [inline pyprecag/tests/test_bandops.py]: should "blud" be "blue"?

File: pyprecag/processing.py [hunk 2]
```diff
@@ -2537,3 +2401,222 @@ def ttest_analysis(points_geodataframe, points_crs, values_raster, out_folder,
...
+                                               output_tif='allyrs_combined.tif')
+
+    with rasterio.open(image_combined, 'r') as src:
+        kwargs = src.meta.copy()
+
+        for i, band in enumerate(src.read(masked=True), start=1):
+            cutoff = np.nanmean(band) + (np.nanmean(band) * (target_percentage / 100.00))
+            band_new = band.copy()
+
+            # classifiy bands
```
Comment (dgobbett) [inline pyprecag/processing.py]: Strictly speaking, this should be slightly different logic as we are testing for >= when the function says >.  And the idea of 'not greater_than' is slightly different from that of 'less than'.

if (greater_than == True) 
    # set to 1 for > cutoff, and 0 otherwise
    band_new[np.where(band > cutoff)] = 1
    band_new[np.where(band <= cutoff)]  = 0
else
    # the 'less than' case 
    #  set to 1 for < cutoff and 0 otherwise
    band_new[np.where(band < cutoff)] = 1
    band_new[np.where(band >= cutoff)]  = 0

File: pyprecag/processing.py [hunk 3]
```diff
@@ -2537,3 +2401,222 @@ def ttest_analysis(points_geodataframe, points_crs, values_raster, out_folder,
...
+    if not config.get_debug_mode() and TEMPDIR in image_combined:
+        os.remove(image_combined)
+
+    return output_tif
+
+
+def persistor_target_probability(upper_raster_files, upper_percentage, upper_probability,
+                                 lower_raster_files, lower_percentage, lower_probability,
+                                 output_tif):
+    """Determine the probability of a performance being exceeded or not being met as described in
```
Comment (dgobbett) [inline pyprecag/processing.py]: As we both realise, this is difficult to summarise for an unfamiliar user. I suggest rewording as follows ... 

"
The "Target probability" method analysis and combines two categories (UPPER and LOWER) of rasters and rules to 
determine which cells meet a higher performance rule (Upper category), or a lower performance rule (Lower 
category) as described in Bramley and Hamilton (2005). This is typically applied to data such as yield maps
with one yield raster per year. 

Each pixel of the output raster is assigned a value as follows:
    1 if the upper category rasters meet the upper category rule, 
    -1 if the lower category rasters meet the lower category rule
    zero otherwise

The upper and lower category rules are defined as follows: 
    Upper: 
        a threshold for each raster calculated from the mean of each raster +/- a specified percentage  
        a proportion of the upper category rasters which must be ABOVE the threshold 
    Lower: 
        a threshold for each raster calculated from the mean of each raster +/- a specified percentage  
        a proportion of the lower category rasters which must be BELOW the threshold

File: pyprecag/processing.py [hunk 4]
```diff
@@ -2537,3 +2401,222 @@ def ttest_analysis(points_geodataframe, points_crs, values_raster, out_folder,
...
+                                    greater_than=True,
+                                    target_percentage=upper_percentage)
+
+    upper_cutoff = (upper_probability / 100.00) * len(upper_raster_files)
+    with rasterio.open(upper_tif, 'r') as src:
+        kwargs = src.meta.copy()
+        band = src.read(masked=True)
+
+        # Greater than test = retuns boolean which can be converted to 0,1
+        data_u = np.where(~band.mask, (band >= upper_cutoff).astype(np.int16), -9999)
```
Comment (dgobbett) [inline pyprecag/processing.py]: Change ">=" to >

File: pyprecag/processing.py [hunk 5]
```diff
@@ -2537,3 +2401,222 @@ def ttest_analysis(points_geodataframe, points_crs, values_raster, out_folder,
...
+                                    temp_file_list[-1],
+                                    greater_than=False,
+                                    target_percentage=lower_percentage)
+
+    lower_cutoff = (lower_probability / 100.00) * len(lower_raster_files)
+    with rasterio.open(lower_tif, 'r') as src:
+        band = src.read(masked=True)
+
+        # Greater than test = retuns boolean which can be converted to 0,1
+        data_l = np.where(~band.mask, np.negative((band >= lower_cutoff).astype(np.int16)), -9999)
```
Comment (dgobbett) [inline pyprecag/processing.py]: Here, since you are using the negative, don't change ">=" to ">"

File: pyprecag/processing.py [hunk 6]
```diff
@@ -2537,3 +2401,222 @@ def ttest_analysis(points_geodataframe, points_crs, values_raster, out_folder,
...
+    lower_tif = persistor_all_years(lower_raster_files,
+                                    temp_file_list[-1],
+                                    greater_than=False,
+                                    target_percentage=lower_percentage)
+
+    lower_cutoff = (lower_probability / 100.00) * len(lower_raster_files)
+    with rasterio.open(lower_tif, 'r') as src:
+        band = src.read(masked=True)
+
+        # Greater than test = retuns boolean which can be converted to 0,1
```
Comment (dgobbett) [inline pyprecag/processing.py]: Comment refers to "Greater"

File: pyprecag/processing.py (hunk 4 above)
Comment (cratcliff) [reply pyprecag/processing.py]: The original code uses the GreaterThanEqual tool for both upper and lower.

File: pyprecag/processing.py (hunk 6 above)
Comment (cratcliff) [reply pyprecag/processing.py]: fixed - deleted line

File: pyprecag/processing.py (hunk 5 above)
Comment (cratcliff) [reply pyprecag/processing.py]: `(band >= lower_cutoff)` returns booleans (0,1) we need -1, 0 so convert to integer and use np.negative and maintain the nodata values. Previously this was done via the GreaterThanEqual tool then a reclass ( 1 1 -1).

File: pyprecag/processing.py (hunk 3 above)
Comment (cratcliff) [reply pyprecag/processing.py]: Updated.

File: pyprecag/processing.py (hunk 2 above)
Comment (cratcliff) [reply pyprecag/processing.py]: This is the same as the original code where we used **'>'** and **'<'**. Ive just converted this to a boolean.
Then it reclassifies using  2 9 0 logic  which as per documentation means assign 0 to values **greater than** 2 and  **less than or equal** to 9 which I have also replicated. 
I can change the variable name if necessary, but this also matches the gui's wording.

File: pyprecag/processing.py (hunk 4 above)
Comment (cratcliff) [reply pyprecag/processing.py]: fixed

**Comment by cratcliff** [comment]
Will fix travisCI at a later date.
